docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md

```diff
@@ -0,0 +1,274 @@
+  because it still has snapshots associated to it.
+* _Source snapshot from which the shallow volume originates is removed while
+  that shallow volume still exists?_
+
+  We need to make sure this doesn't happen and some book-keeping
+  is necessary. Ideally we could employ some kind of reference counting.
+
+#### Reference counting for shallow volumes
+
+As mentioned above, this is to protect shallow volumes, should their source
+snapshot be requested for deletion.
+
+When creating a volume snapshot, a new RADOS object would be created for that
+snapshot. It would contain a single signed integer value that would describe
+the number of references the backing subvolume snapshot has. It would be
+initialized to one, where the CSI snapshot itself is the first reference to the
+backing snapshot. Each subsequent shallow volume creation would increment the
+reference counter, and each shallow volume deletion would decrement it.
+Calling `DeleteSnapshot` would decrement the reference counter as well.
+
+The subvolume snapshot would be removed from the Ceph cluster only once the
+reference count reaches zero. Note that this behavior would permit calling
+`DeleteSnapshot` even if it is still referenced by shallow volumes.
+
+* `DeleteSnapshot`:
+  * `refcount == 0` or the respective _refcount_ RADOS object doesn't exist:
+    delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+* `DeleteVolume`:
+  * `refcount == 0`: delete the backing snapshot too.
+  * `refcount > 0`: keep the backing snapshot.
+
+To enable creating shallow volumes from snapshots that were provisioned by
+older versions of cephfs-csi (i.e. before this feature is introduced),
+`CreateVolume` for shallow volumes would also create the _refcount_ object in
+case it's missing. It would be initialized to two: the source snapshot and the
+newly created shallow volume.
+
+##### Concurrent access to _refcount_ objects
+
+We can't do operations on integers in RADOS objects atomically, which means we
+need to do all synchronization ourselves when modifying the value of the counter.
+
+From a single master perspective, concurrent access to _refcount_ can be
+protected by mutex locks and is easy to implement. `cephfs.ControllerServer`
+already offers `VolumeLocks`, `SnapshotLocks` and `OperationLocks`.
+
+Mutexes won't cover situations where multiple masters may be active though
+(during/after leader election), and modifying _refcount_ would still race in
+this case. Misbehaving master that lost leadership would still be able to
+access resources on the Ceph cluster. Racing on _refcount_ could then cause
+leaking the _refcount_ object itself and/or the subvolume snapshot, or the
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
```

Comment:
Add a note in the document when we implement this one. To use this feature and to avoid issues run Deployment with replica `1`

Reply:
https://github.com/ceph/ceph-csi/pull/2148/commits/b02a288ecad4694cfaf87d770f828c7c9191584a